Improve schema generation for return object

hideya opened the pull request:
Fix https://github.com/strongloop/loopback-swagger/issues/27

Updated loopback-swagger has been tested by generating Swagger Spec of [this test API server](https://docs.strongloop.com/display/public/LB/Create+a+simple+API), feeding it to [swagger-codegen](https://github.com/swagger-api/swagger-codegen) and writing a simple client in Swift using the swagger-codegen-generated SDK.  I tested invocation of `PersistedModel.count`, which demonstrated the issue, in the simple test client and made sure this fix works.

@bajtos would you please review the changes?

hideya commented:
**Additional Note:**
This fix brings up another issue of `ReadableStream` not being defined in generated Swagger Spec.  After this fix, generated Spec starts including lines like the following, although `ReadableStream` is not legal as the schema property type:
```yaml
  /Users/change-stream:
      ...
      responses:
        '200':
          description: Request was successful
          schema:
            type: object
            required:
              - changes
            properties:
              changes:
                type: ReadableStream
```
FYI, swagger-codegen handles this gracefully (emits warnings but doesn't abort) and generates a usable SDK (although `/Users/change-stream` part doesn't work).

hideya commented on lib/specgen/route-helper.js:
```diff
@@ -96,19 +96,25 @@ var routeHelper = module.exports = {
-    var def = { type: 'object' };
-    return schemaBuilder.buildFromLoopBackType(def, typeRegistry);
+    // Construct scheme for the return object
+    var schema = { type: 'object' };
```
As far as I can tell, we don't need to invoke `schemaBuilder.buildFromLoopBackType()` here just for `{ type: 'object' }`, but leaving this comment since it was invoked originally and thought I may be missing something.

bajtos commented on lib/specgen/route-helper.js:
```diff
@@ -96,19 +96,25 @@ var routeHelper = module.exports = {
+    schema.properties = {};
+
+    routeReturns.forEach(function(ret) {
+      var propName = ret.arg;
```
I believe we support both `arg` and `name` for specifying argument name, see https://github.com/hideya/loopback-swagger/blob/d1b01a7d2acff76b7af5e010e64ee550df62c8fe/lib/specgen/route-helper.js#L219 and https://github.com/strongloop/strong-remoting/blob/a6a75fc86a429458294168d2751df9a013b85da3/lib/shared-method.js#L299.

bajtos replied:
Please add a unit-test for `ret.name` too.

hideya replied:
Thanks for pointing those out!  Both are done in my local copy.

bajtos commented on lib/specgen/route-helper.js:
```diff
@@ -96,19 +96,25 @@ var routeHelper = module.exports = {
+
+    routeReturns.forEach(function(ret) {
+      var propName = ret.arg;
+      var propType = ret.type;
```
What if `ret.type` is a constructor function, e.g. `MyModel` or `String`? Or an array like `['string']` or `[String]`? I think you should call `buildFromLoopBackType` here to handle these different ways of specifying a type.

Also please add unit-tests to verify the following:
 - a built-in constructor - `returns: { type: String }`
 - a known model class - `returns: { type: ModelWeKnow }`
 - a known model name - `returns: { type: 'ModelWeKnow' }`
 - an array type, at least `['string']`, possibly some/all of the three above cases too

bajtos replied:
Ha, this is more complex than I initially thought. To support model types, we need to reference them via `typeRegistry`, to make sure their definitions are included in the swagger too.

Let's leave that out of scope of this pull request so that we can get something landed soon.

So when you encounter a type (name, constructor) that is not a known built-in/primitive type, please print a warning and emit `type: 'object', additionalProperties: true` instead.

I hope I didn't confuse you, feel free to ping me on Skype so that we can clarify.

hideya replied:
> To support model types, we need to reference them via typeRegistry, to make sure their definitions are included in the swagger too.

Would the following code possibly work ok, do you think?
```js
    var schema = { type: 'object' };
    schema.required = [];
    schema.properties = {};

    routeReturns.forEach(function(ret) {
      var propName = ret.name || ret.arg;
      if (ret.model) {
        schema.properties[propName] = { $ref: typeRegistry.reference(ret.model) };
      } else {
        var propType = schemaBuilder.buildFromLoopBackType(ret, typeRegistry);
        schema.properties[propName] = { type: propType };
      }
      schema.required.push(propName); // TO BE FIXED
    });
```

hideya replied:
> Also please add unit-tests to verify the following:

Yes, I've added them to my local workspace.

> when you encounter a type (name, constructor) that is not a known built-in/primitive type,

Would you kindly tell me what is the easiest way to determine this?

bajtos replied:
> > when you encounter a type (name, constructor) that is not a known built-in/primitive type,

> Would you kindly tell me what is the easiest way to determine this?

@hideya Ha, that's a good question!

A primitive type can be detected by `schemaBuilder.isPrimitiveType`, see 
https://github.com/strongloop/loopback-swagger/blob/cc10105fdc7780e05bca37b9070b48aa99a7cb3a/lib/specgen/schema-builder.js#L177-L179

A known custom type (definition) can be detected by `typeRegistry.isDefined`, see https://github.com/strongloop/loopback-swagger/blob/cc10105fdc7780e05bca37b9070b48aa99a7cb3a/lib/specgen/type-registry.js#L82. 

I think the check for custom types (definitions) can be omitted from from this patch, to keep things simple.

I am proposing this: If `isPrimitiveType` returns false, then use `type:object` instead and print a warning via `console.warn`, so that people let us know when they actually run into this missing bit.

You also need to run `schemaBuilder.getLdlTypeName` first to get a string representation of the raw type provided in remoting metadata, i.e. convert string/constructor/array into a string value that `isPrimitiveType` understands.

```js
var typeString = schemaBuilder.getLdlTypeName(ret.type);
if (schemaBuilder.isPrimitiveType(typeString)) {
  // use the value in typeString for Swagger/JSON schema
} else {
  // console.warn(...)
  // use "object" for Swagger/JSON schema
}
```

Thoughts?

hideya replied:
@bajtos sounds good to me!  I implemented in the following way:
```js
      var idlType = schemaBuilder.getLdlTypeName(ret.type);
      if (Array.isArray(idlType) || schemaBuilder.isPrimitiveType(idlType)) {
        schema.properties[propName] = schemaBuilder.buildFromLoopBackType(ret, typeRegistry);
      } else {
        console.warn('Swagger: skipping unknown type %j found in route %j', idlType, route);
      }
```
and by trying it out, found that `/DataSourceDefinitions/:id/testConnection` yields the following warning:
```
Swagger: skipping unknown type "Error" found in route {"verb":"get","path":"/DataSourceDefinitions/:id/testConnection","documented":true,"method":"DataSourceDefinition.prototype.testConnection","returns":[{"arg":"status","type":"boolean"},{"arg":"error","type":"Error"}]}
```
this seems originated by https://github.com/strongloop/loopback-workspace/blob/6b537f0657cc138430e5fe72f441e95b3b04e510/common/models/data-source-definition.js#L76
Is this something needs to be fixed in loopback-workspace?

bajtos replied:
> this seems originated by https://github.com/strongloop/loopback-workspace/blob/6b537f0657cc138430e5fe72f441e95b3b04e510/common/models/data-source-definition.js#L76
> Is this something needs to be fixed in loopback-workspace?

Yes, this looks like a bug in loopback-workspace to me: method's remoting metadata references a type (`Error`) that is not defined inside loopback core and loopback-workspace does not provide definition of this type either.

bajtos replied:
I created an issue in loopback-workspace to fix this problem: https://github.com/strongloop/loopback-workspace/issues/269

bajtos commented on lib/specgen/route-helper.js:
```diff
@@ -96,19 +96,25 @@ var routeHelper = module.exports = {
+    routeReturns.forEach(function(ret) {
+      var propName = ret.arg;
+      var propType = ret.type;
+      schema.required.push(propName);
```
I have mixed feelings about specifying all arguments as required. I don't think that's always true, consider a function that calls either `cb(null, 'foo', 'bar')` or `cb(null, 'baz')` depending on the input.

Is this needed for the Swift codegen? 

I would prefer to check `ret.required` to decide whether a return arg is required or optional. I am wondering if we could re-use the code in `model-helper.js` which already takes care of that?

hideya replied:
Maybe I'm totally mistaken but I don't find a way to get `required` info for return arguments...
For example, I can get the following info from Strong Remoting Route object:
```json
{
  "verb":"get",
  "path":"/Products/count",
  "description":"Count instances of the model matched by where from the data source.",
  "documented":true,
  "method":"Product.count",
  "accepts":[{"arg":"where","type":"object", "description":"Criteria to match model instances"}],
  "returns":[{"arg":"count", "type":"number"}]
}
```
but I don't see any `required` info for the properties of the return object.
`prop.required` checks in `model-helpers.js` seem to me just dealing with properties of models.
Would you tell me where I lost...?

bajtos replied:
Interesting. I think nobody was using `required` for `returns` args so far, because strong-remoting does not interpret this flag, even if the documentation allows it (see https://docs.strongloop.com/display/LB/Remote+methods#Remotemethods-Argumentdescriptions). That's why you don't see any `required` flag in `returns` metadata above.

I am proposing to honour the `required` flag in loopback-swagger.

I was just about to propose to modify remoting metadata in LoopBack to specify `required:true` for the `count` property, but then I realised that not all connectors support this feature, i.e. not all connectors report the number of deleted records, in which case `count` is not set.

However, the case of `/Products/:id/exists` described in your previous comment is different, I think that method always returns true or false, and thus it is ok to add `required:true`.

Thoughts?

bajtos replied:
In case it's not clear: the `required` flag is optional and defaults to `false`. That's why `"returns":[{"arg":"exists", "type":"boolean"}]` is a valid description.

hideya replied:
@bajtos ok, sounds good.  I'll update this PR accordingly.  As you noticed, swagger-codegen folks are preparing a PR for this change: **For "type: object" without any property or $ref, we should map it as generic object in the respective language** https://github.com/swagger-api/swagger-codegen/issues/2093#issuecomment-191168537
So, those optional return arguments will become accessible.

bajtos replied:
> (NOTE: the return argument for /<PersistedModel>/:id/exists becomes inaccessible until required: true is specified for it, or swagger-codegen folks merges the above mentioned PR)

Ouch. Are you sure that swagger codegen does not emit Swift property getters for optional properties? How difficult would it be to fix that? 

On the other hand, I think it makes perfect sense to add `required:true` to remoting metadata for this particular method, so it's not a big deal how Swift codegen handles optional properties. @hideya Do you have time to submit the PR fixing remoting metadata in loopback yourself?

bajtos commented:
@hideya as for `ReadableStream`, how difficult would it be to detect this particular case and emit `{ schema: { type: 'file' } }` instead?

See Swagger spec: http://swagger.io/specification/#responseObject
> As an extension to the Schema Object, its root type value may also be "file". This SHOULD be accompanied by a relevant produces mime-type.

By default, we produce `application/json; boundary=NL` as the content (mime) type, see https://github.com/strongloop/strong-remoting/pull/216

How will the codegen client API look like?

hideya commented:
> @hideya as for ReadableStream, how difficult would it be to detect this particular case and emit { schema: { type: 'file' } } instead?

It is quite easy and I've done it.  Generated Swift method looks like the following, and gives a URL back.
```swift
public class func userCreateChangeStreamGetUsersChangeStream(options options: String?, completion: ((data: NSURL?, error: ErrorType?) -> Void)) {
```

FYI, the yaml for this method looks like the following:
```yaml
  /Users/change-stream:
    post:
      tags:
        - User
      summary: Create a change stream.
      description: !<tag:yaml.org,2002:js/undefined> ''
      operationId: User.createChangeStream__post_Users_change-stream
      parameters:
        - name: options
          in: formData
          description: !<tag:yaml.org,2002:js/undefined> ''
          required: false
          type: string
          format: JSON
      responses:
        '200':
          description: Request was successful
          schema:
            type: file
      deprecated: false
```

hideya commented:
@bajtos I've updated the PR to reflect your review.  There are following two remaining points which I haven't dealt with:

1.  when you encounter a type (name, constructor) that is not a known built-in/primitive type, please print a warning and emit `type: 'object', additionalProperties: true` instead.
*  check `ret.required` to decide whether a return arg is required or optional -- https://github.com/strongloop/loopback-swagger/pull/28#discussion_r54258385

As to 1., would you kindly tell me what is the easiest way to determine this condition? **a type (name, constructor) that is not a known built-in/primitive type**.   I checked out `type-registry.js` for some hits but no luck...

As to 2., I'll cut and paste my question below for your convenience:
***
Maybe I'm totally mistaken but I don't find a way to get `required` info for return arguments...
For example, I can get info like the following from a Strong Remoting Route object:
```json
{
  "verb":"get",
  "path":"/Products/:id/exists",
  "description":"Check whether a model instance exists in the data source.",
  "documented":true,
  "method":"Product.exists",
  "accepts":[{"arg":"id", "type":"any", "description":"Model id", "required":true}],
  "returns":[{"arg":"exists", "type":"boolean"}]
}
```
but I don't see any `required` info for the properties of the return object (I see `required` property in `"accepts"` but haven't seen any in `"returns"`).
`prop.required` checks in `model-helpers.js` seem to me dealing with properties of models.
Could you tell me where I lost...?

bajtos commented:
@raymondfeng @ritch @STRML  in my https://github.com/strongloop/loopback-swagger/pull/28#discussion_r54874551 above, I am proposing to start interpreting `required:true` flag in *returns* arguments too. My intention is to interpret `required` as whether the argument value will be defined. I don't want to interpret `required:false` as a flag saying that the argument may be omitted from callback args.

```
{ returns: [
  { arg: 'arg1', required: false },
  { arg: 'arg2', required: true },
]}

function fn(cb) {
  cb(null, undefined, 'arg2');
  // ^^^ this is valid

  cb(null, 'arg2');
  // ^^ this is not valid
  // it will be interpreted as cb(null, 'arg2', undefined)
  // i.e. arg1='arg2', arg2=undefined
}
```

I am concerned that this may be a bit confusing, but on the other hand I don't want start dealing with complex edge cases that would need to be solved if we started to treat callback arguments as truly optional.

My proposal uses the `required` flag only for the convenience of Swagger consumers and possible validators of responses, but does not modify any runtime behaviour.

@raymondfeng @ritch @STRML thoughts?

hideya commented:
@bajtos I've updated this PR once again.  I think I addressed all the points you brought up.  Would you please review the updates?

hideya commented:
@bajtos I think I addressed all the points you mentioned and updated the PR.  Could you please review the changes?  Also, could you quickly point out the reasons why the three CI tests are failing?

bajtos commented on lib/specgen/route-helper.js:
```diff
@@ -100,6 +100,8 @@ var routeHelper = module.exports = {
       }
       return schemaBuilder.buildFromLoopBackType(routeReturns[0], typeRegistry);
+    } else if (routeReturns.length === 1 && routeReturns[0].type === 'ReadableStream') {
+      return { type: 'file' };
```
@hideya Please add a unit-test for this case.

hideya replied:
@bajtos done :-)

bajtos commented on lib/specgen/route-helper.js:
```diff
@@ -108,10 +110,25 @@ var routeHelper = module.exports = {
+      var genericIdlType = { type: 'object' };
+      while (Array.isArray(itemIdlType)) {
+        itemIdlType = schemaBuilder.getLdlTypeName(itemIdlType[0]);
+        genericIdlType = { type: 'array', items: genericIdlType };
```
@hideya I think `items` should be set to `itemIdlType`. Am I missing something?

bajtos replied:
Oh, I think I am starting to understand this code. `genericIdlType` is used in case when we have an array of non-primitive types. It makes me wonder why is this needed? I would expect that `buildFromLoopBackType` takes care of that.

hideya replied:
@bajtos I've added the following comment and a test case for a nested array of an unknown type.
```
     // Note that we cannot simply use buildFromLoopBackType since it converts unknown type to
     // '$ref': '#/definitions/UnknownType', whereas we decided to emit 'type: object' for such a case.
     // See https://github.com/strongloop/loopback-swagger/pull/28#discussion_r54873911
     // The following code is needed to take care of a nested array of an unknown type.
```
Hope this make sense.

bajtos commented on lib/specgen/route-helper.js:
```diff
@@ -108,10 +110,25 @@ var routeHelper = module.exports = {
+      if (schemaBuilder.isPrimitiveType(itemIdlType)) {
+        schema.properties[propName] = schemaBuilder.buildFromLoopBackType(ret, typeRegistry);
+      } else {
+        console.warn('Swagger: skipping unknown type %j found in route %j', itemIdlType, route);
```
This warning looks misleading to me, because the next line says "Temporarily use `object` as the (inner item) type", i.e. the unknown type is not really ignored.

hideya replied:
I've fixed it to the following:
```js
console.warn('Swagger: temporarily use `object` instead of unknown type %j found in route %j', itemIdlType, route);
```

bajtos commented on test/specgen/route-helper.test.js:
```diff
@@ -4,25 +4,61 @@ var routeHelper = require('../../lib/specgen/route-helper');
     });
-    expect(getResponseMessage(entry.operation))
+    var responseMessage = getResponseMessage(entry.operation);
+    ((responseMessage || {}).schema || {}).required.sort(); // sort the array for the comparison below
```
If `schema` is not defined, then this resolves to `({}).required.sort()`, which throws `TypeError: Cannot read property 'sort' of undefined`.

hideya replied:
Actually, it was failing when `required` was not defined.  I've fixed the above line into the following:
```js
(((responseMessage || {}).schema || {}).required || {}).sort();  // sort the array for the comparison below

```

bajtos replied:
I think the last bit should be `.required || []`, i.e. an empty array instead of an empty object?

hideya replied:
You are right!  I've fixed this one too.  Thank you for pointing it out!

bajtos commented:
@hideya reviewed, see my comments above.

hideya commented:
@bajtos I think I addressed all the points you've made in the comments.  Please let me know how the PR looks like.  If it looks good, I'll clean up the commit history.

bajtos commented on lib/specgen/route-helper.js:
```diff
@@ -122,8 +126,7 @@ var routeHelper = module.exports = {
       } else {
-        console.warn('Swagger: skipping unknown type %j found in route %j', itemIdlType, route);
-        // Temporarily use `object` as the (inner item) type
+        console.warn('Swagger: temporarily use `object` instead of unknown type %j found in route %j', itemIdlType, route);
```
Please reword to `Swagger: temporarily using...` (instead of `...use...`).

hideya replied:
Fixed! Thanks for pointing it out.

bajtos commented:
One more minor item to fix (see above) and this is good to go :+1: @hideya please squash the commits right after you fix it, so that we don't have to go through another roundtrip here.

hideya commented:
@bajtos I've applied the fixes and squashed the commit history.  Hope this PR looks fine by now.  I can still spare short time if needed during weekends.  Please don't hesitate to let me know if I can help!

bajtos commented:
@hideya landed via  b23e08c, thank you for the contribution!

hideya commented:
@bajtos thank you for letting me contribute to this great project!  Appreciated!
